feat(optimizer): Implement CountWindowElimination logic

tennisleng opened the pull request:
Splitting #20076: Core logic and tests.

tennisleng committed:
feat(optimizer): Scaffold CountWindowElimination rule

tennisleng committed:
feat(optimizer): Implement CountWindowElimination logic

lnkuiper reviewed (changes requested):
Thanks for the PR. It seems that with this PR, PR https://github.com/duckdb/duckdb/pull/20080 is now redundant. If so, could you close it? I don't see why this needs to be split into two PRs.

Also, could you implement the feedback given by @hawkfish in your initial PR?
https://github.com/duckdb/duckdb/pull/20076#pullrequestreview-3549598632

tennisleng committed:
refactor: Address hawkfish feedback on CountWindowElimination

- Rename class CountWindowElimination -> WindowSelfJoinOptimizer
  (for future extensibility per VLDB research on window-to-self-join)
- Rename inner class TableRebinder -> CountWindowTableRebinder
  (for amalgamation build safety)
- Use COMPARE_NOT_DISTINCT_FROM instead of COMPARE_EQUAL for join
  conditions (to handle NULL partition keys correctly)

hawkfish commented on src/optimizer/count_window_elimination.cpp:
```diff
@@ -216,7 +216,7 @@ unique_ptr<LogicalOperator> CountWindowElimination::OptimizeInternal(unique_ptr<
 			for (size_t i = 0; i < w_expr.partitions.size(); ++i) {
 				JoinCondition cond;
-				cond.comparison = ExpressionType::COMPARE_EQUAL;
+				cond.comparison = ExpressionType::COMPARE_NOT_DISTINCT_FROM;
```
Can we also add a test for this?

tennisleng committed:
test: Add NULL partition key test for COMPARE_NOT_DISTINCT_FROM

This test verifies that the optimizer correctly uses COMPARE_NOT_DISTINCT_FROM
instead of COMPARE_EQUAL for join conditions, ensuring that partition keys
containing NULL values are properly matched during the window-to-semi-join
transformation.

Addresses review comment from @hawkfish on PR #20081.

lnkuiper reviewed (changes requested):
Thanks for the changes! I have some more feedback. Maybe @hawkfish could also have one last look?

hawkfish reviewed (commented):
The reason I'm asking to rename the functionality is that with some follow on work this PR could address #20075 and #6042. In particular we could generalise it to multiple aggregates that share the same partitioning and import the aggregates. We could also remove the predicate check as the predicate will get pushed down through the join by the optimiser. But that is for a later pass.

hawkfish reviewed (approved):
I can rename things after the merge

tennisleng committed:
fix: Address CI failures - format and missing binder.hpp include

- Fixed clang-format issue on line 86 (parameter alignment)
- Added missing #include "duckdb/planner/binder.hpp" for Windows build (C2027 error)
